pallets/loans-ref/src/lib.rs

```diff
@@ -0,0 +1,775 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub mod types;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+mod pallet {
+	use cfg_primitives::Moment;
+	use cfg_traits::{ops::EnsureAdd, InterestAccrual, Permissions, PoolInspect, PoolReserve};
+	use cfg_types::{
+		adjustments::Adjustment,
+		permissions::{PermissionScope, PoolRole, Role},
+	};
+	use frame_support::{
+		pallet_prelude::*,
+		traits::{
+			tokens::{
+				self,
+				nonfungibles::{Inspect, Mutate, Transfer},
+			},
+			UnixTime,
+		},
+		transactional, PalletError, StorageHasher,
+	};
+	use frame_system::pallet_prelude::*;
+	use scale_info::TypeInfo;
+	use sp_arithmetic::FixedPointNumber;
+	use sp_runtime::{
+		traits::{BadOrigin, Zero},
+		FixedPointOperand,
+	};
+	use types::{
+		ActiveLoan, AssetOf, ClosedLoan, CreatedLoan, LoanInfo, LoanRestrictions,
+		PortfolioValuation, PortfolioValuationUpdateType, RepaymentSchedule, ValuationMethod,
+		WriteOffState, WriteOffStatus,
+	};
+
+	use super::*;
+
+	type PoolIdOf<T> = <<T as Config>::Pool as PoolInspect<
+		<T as frame_system::Config>::AccountId,
+		<T as Config>::CurrencyId,
+	>>::PoolId;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Identify a curreny.
+		type CurrencyId: Parameter + Copy + MaxEncodedLen;
```

Comment:
That would be great. I already tried to get rid of these redundant config parameters, but I couldn't because I needed a way to add some bounds to them. For example, `CollectionIdOf<T>` does not have the required bound `Parameter` that this library needs because it is used as a parameter in create()` method. This applies to `ItemId`, `CollectionId`, `Balance` and `CurrencyId`. 🥲 

I know that I can give an explicit list of bounds each time I use `T, but it becomes too much verbose in many places, and the bounds are propagated. Not sure if it is worth it 🤔

Reply:
What about bounding the trait? 

But I think you would need to add all those bounds to the T used places too till the compiler gets smart enough.

Reply:
Yes... I need to add all trait bounds of all T occurrences 😞. Not sure why realizing this is "difficult" for the compiler.